spotify_player/src/event/window.rs

```diff
@@ -288,6 +371,29 @@ pub fn handle_command_for_track_list_window(
     Ok(true)
 }
 
+pub fn handle_action_for_artist_window(
```

Comment:
this and `handle_action_for_{album/track/playlist}_window` basically share same logic, which might be simplified with macros or generics. It looks good as-is for me now because the PR is kinda large already. I can explore different options after this lands.

Reply:
macro can be relatively simple to add [with my other suggestion](https://github.com/aome510/spotify-player/pull/471#discussion_r1649788012). However, generic might be a bit tricky to get it working

Reply:
This would work

```rust
pub trait ActionHandler {
    fn handle_action_for_window(
        action: Action,
        items: Vec<&Self>,
        data: &DataReadGuard,
        ui: &mut UIStateGuard,
        client_pub: &flume::Sender<ClientRequest>,
    ) -> Result<bool>;
}

#[macro_export]
macro_rules! impl_action_handler {
    ( $context_type:ident ) => {
        impl ActionHandler for $context_type {
            fn handle_action_for_window(
                action: Action,
                items: Vec<&Self>,
                data: &DataReadGuard,
                ui: &mut UIStateGuard,
                client_pub: &flume::Sender<ClientRequest>,
            ) -> Result<bool> {
                let id = ui.current_page_mut().selected().unwrap_or_default();
                if id >= items.len() {
                    return Ok(false);
                }

                handle_action_in_context(
                    action,
                    ActionContext::$context_type(items[id].clone()),
                    client_pub,
                    data,
                    ui,
                )?;

                Ok(true)
            }
        }
    };
}

impl_action_handler!(Playlist);

Playlist::handle_action_for_window(...)
```

Reply:
I've implemented generic approach in https://github.com/aome510/spotify-player/pull/471/commits/6a8bd894f1dffe667dde76bdb596fc181977e327. It's simpler than I previously thought. I would avoid using macros if possible because formatting (and other stuff) might not work well in the macro block.

Reply:
Ah yes, this is a much more elegant solution